Suggested fixes for a few possible trivial mistakes in L1Trigger/L1TTrackMatch

perrotta opened the pull request:
#### PR description:

Starting from some sparse compilation and static analyzer warnings, I found a few apparently trivial mistakes (bugs?) in the following plugins:
- L1Trigger/L1TTrackMatch/plugins/L1TkEmParticleProducer.cc
- L1Trigger/L1TTrackMatch/plugins/L1TkFastVertexProducer.cc

The fixes suggested in this attempt PR are able to get rid of the warnings. But I suspect that they can only hide the real problem, which would require a more informed bug fix campaign there.

@cecilecaillol and @cms-sw/l1-l2 could you please have a look?

#### PR validation:

It builds
As such, no changes are expected in outputs. But I don't think the first implementation can be considered the final fix

perrotta committed:
Suggested fixes for a few possible trivial mistakes

perrotta commented on L1Trigger/L1TTrackMatch/plugins/L1TkEmParticleProducer.cc:
```diff
@@ -266,7 +266,7 @@ float L1TkEmParticleProducer::CorrectedEta(float eta, float zv) const {
   } else {
     if (theta > 0)
       delta = ZEcal;
-    if (theta < 0)
```
A few lines above `theta` is forced to be positive.
Therefore, the delta for the negative endcap will never be set correctly
This is a bug, which is **not** fixed by this change (it only avoids the static analyzer to complain because tha case `theta=0` would let `delta` uninitialized in the following)

perrotta replied:
(Perhaps just using `eta` instead of `theta` in the `if` clause may be enough to fix. Still, the whole code, at least for this method, is written in a quite inefficient way)

cecilecaillol replied:
@rekovic I believe you are the original author of this code? Can you check this comment?

perrotta replied:
> @rekovic @cecilecaillol Do you agree on the proposed change?

Since @cecilecaillol has already signed this PR, I would say yes...

cecilecaillol commented:
@perrotta Thank you! 
@emacdonald16 Could you please have a look at this correction PR?

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33794/22767

- This PR adds an extra 16KB to repository

cmsbuild commented:
A new Pull Request was created by  @perrotta  for master.

It involves the following packages:

L1Trigger/L1TTrackMatch


@cmsbuild, @rekovic, @srimanob, @cecilecaillol, @kpedro88 can you please review it and eventually sign? Thanks.
@Martin-Grunewald this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

perrotta commented on L1Trigger/L1TTrackMatch/plugins/L1TkFastVertexProducer.cc:
```diff
@@ -309,25 +309,10 @@ void L1TkFastVertexProducer::produce(edm::StreamID, edm::Event& iEvent, const ed
+  float sigma_max;
+  int imax;
   int nb = htmp.GetNbinsX();
-  for (int i = 2; i <= nb - 1; i++) {
```
All this loop compute uselessly variables which are recompute anyhow in the following one.
The trivial fix wold be to remove this loop as a whole, as suggested here: but perhaps the author of the code hade something else in mind, so better to have an expert look.

emacdonald16 replied:
You are correct that this loop is unnecessary, and removing it would be fine. However, this is the old vertex producer that we're migrating away from, and will soon be obsolete. The newer producer we are using going forward is L1Trigger/VertexFinder

emacdonald16 replied:
Let's remove the loop now, then, since I'm not entirely sure when this will become strictly obsolete.

perrotta committed:
fix: theta -> eta

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33794/22790

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #33794 was updated. @cmsbuild, @rekovic, @srimanob, @cecilecaillol, @kpedro88 can you please check and sign again.

perrotta commented:
please test
(let try testing with the most obvious bug fix)

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e36b77/15226/summary.html
**COMMIT**: 8b97139d41a46046b066ce520f8747927ad15d46
**CMSSW**: CMSSW_12_0_X_2021-05-20-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/33794/15226/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2650486
 - DQMHistoTests: Total failures: 12
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2650451
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 36 files compared)
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 155 log files, 37 edm output root files, 37 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-05-20-2300+e36b77/42899/triggerResults): no differences found

perrotta commented:
@cecilecaillol @rekovic : could anybody tell me if L1TkEmParticleProducer is used anywhere? It apparently enters L1Trigger/Configuration/python/SimL1Emulator_cff.py, but I am puzzled by the fact that the bug fix here does not reflect in the Phase2 outputs anywhere. Please advise.

perrotta committed:
Code improvents for L1TkEmParticleProducer

perrotta commented:
> @cecilecaillol @rekovic : could anybody tell me if L1TkEmParticleProducer is used anywhere? It apparently enters L1Trigger/Configuration/python/SimL1Emulator_cff.py, but I am puzzled by the fact that the bug fix here does not reflect in the Phase2 outputs anywhere. Please advise.

I investigated it myself. Indeed the plugin is used in step2 by the Phase2 L1 emulator with two instances, namely `L1TkPhotonsCrystal` and `L1TkPhotonsHGC.` However, those modules are run with `PrimaryVtxConstraint = false`, and therefore the bug does not enter in the selection, which explains why there are no differences visible in the test outputs. Moreover, the bug only affects one half of the L1TkEm in the endcaps, that is a tiny fraction (if any) of the entries in the workflows tested.

Since the bug is there, and the plugin is actually used, I'd say that it has to be fixed. While looking at it, I also implemented a few quite simple code changes meant to improve the cpu performance (even though the two modules take already a very small time, and the time improvement achieved does not propagate in a visible way to the overall performance); moreover, since I was there, I also fixed the `fillDescriptions` method, which was previously doing nothing for the parameter validation, and adapted accordingly the explicit python configurations used in the workflows.

Because of all what above, I remove now the "RFC" from the title and ask @cms-sw/l1-l2 to evaluate and possibly sign it for merging.

perrotta commented:
type bug fix

(where the bug fix is only in the determination of the sign of the endcaps inside L1TkEmParticleProducer::CorrectedEta, all the rest being some code improvement and cleanings)

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33794/22860

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #33794 was updated. @cmsbuild, @rekovic, @srimanob, @cecilecaillol, @kpedro88 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e36b77/15304/summary.html
**COMMIT**: 73660445ba1159c035079fbb79f3e64cf26b0d83
**CMSSW**: CMSSW_12_0_X_2021-05-25-1700/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/33794/15304/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2650486
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2650457
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 36 files compared)
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 155 log files, 37 edm output root files, 37 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-05-25-1700+e36b77/42959/triggerResults): no differences found

cecilecaillol commented:
+l1

srimanob commented on L1Trigger/L1TTrackMatch/plugins/L1TkEmParticleProducer.cc:
```diff
@@ -285,8 +272,21 @@ void L1TkEmParticleProducer::fillDescriptions(edm::ConfigurationDescriptions& de
   //The following says we do not know what parameters are allowed so do no validation
```
I think the comment can be deleted as `desc.setUnknown();` is done.

srimanob commented:
+Upgrade

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented:
+1
